sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala

```diff
@@ -270,12 +274,12 @@ case class FileSourceScanExec(
         "PushedFilters" -> seqToString(pushedDownFilters),
         "DataFilters" -> seqToString(dataFilters),
         "Location" -> locationDesc)
-    val withOptPartitionCount =
-      relation.partitionSchemaOption.map { _ =>
-        metadata + ("PartitionCount" -> selectedPartitions.size.toString)
-      } getOrElse {
-        metadata
-      }
+    val withOptPartitionCount = if (relation.partitionSchemaOption.isDefined &&
+      !partitionOnlyAvailableAtRuntime) {
+      metadata + ("PartitionCount" -> selectedPartitions.size.toString)
+    } else {
+      metadata
```

Comment:
I am wondering about adding here `"PartitionCount" -> "unknown"` or something which shows that we are reading partitioned data but we don't know how many of them. On the other side maybe people trust this field to be a numeric one so I am not sure about the best thing to do honestly. @cloud-fan @viirya thoughts?

Reply:
I think current way is good. We don't need to include unknown metadata.